src/equations/svaerd_kalisch_1d.jl

```diff
@@ -174,129 +182,188 @@ function create_cache(mesh, equations::SvaerdKalischEquations1D,
+            mul!(deta, D1.minus, tmp1)
+            mul!(deta, D1_central, hv, -1.0, 1.0)
         else
-            @error "unknown type of first derivative operator: $(typeof(solver.D1))"
+            @error "unknown type of first derivative operator: $(typeof(D1))"
         end
     end
 
     # split form
     @trixi_timeit timer() "dv hyperbolic" begin
-        D1_hv = D1_central * hv
-        D1_hv2 = D1_central * (hv .* v)
-        D1_gamma_hat_D2_v = D1_central * (gamma_hat .* (solver.D2 * v))
-        D2_gamma_hat_D1_v = solver.D2 * (gamma_hat .* D1v)
-        @. dv = -(0.5 * (D1_hv2 + hv * D1v - v * D1_hv) +
-                  equations.gravity * h * D1eta +
-                  0.5 * (vD1y - D1vy - yD1v) -
-                  0.5 * D1_gamma_hat_D2_v -
-                  0.5 * D2_gamma_hat_D1_v)
+        mul!(h_v_x, D1_central, hv)
+        @.. tmp1 = hv * v
+        mul!(hv2_x, D1_central, tmp1)
+        mul!(v_xx, solver.D2, v)
+        @.. tmp1 = gamma_hat * v_xx
+        mul!(gamma_v_xx_x, D1_central, tmp1)
+        @.. tmp1 = gamma_hat * v_x
+        mul!(gamma_v_x_xx, solver.D2, tmp1)
+        @.. dv = -(0.5 * (hv2_x + hv * v_x - v * h_v_x) +
+                   g * h * eta_x +
+                   0.5 * (v_y_x - vy_x - y_v_x) -
+                   0.5 * gamma_v_xx_x -
+                   0.5 * gamma_v_x_xx)
     end
 
     # no split form
     #     dv[:] = -(D1_central * (hv .* v) - v .* (D1_central * hv)+
-    #               equations.gravity * h .* D1eta +
-    #               vD1y - D1vy -
-    #               0.5 * D1_central * (gamma_hat .* (solver.D2 * v)) -
-    #               0.5 * solver.D2 * (gamma_hat .* D1v))
+    #               g * h .* eta_x +
+    #               vy_x - v_y_c -
+    #               0.5 * gamma_v_xx_x -
+    #               0.5 * gamma_v_x_xx)
 
     @trixi_timeit timer() "source terms" calc_sources!(dq, q, t, source_terms, equations,
                                                        solver)
-    @trixi_timeit timer() "dv elliptic" begin
-        # decompose M * (h - D1betaD1) because it is guaranteed to be symmetric and pos. def.,
-        # while (h - D1betaD1) is not necessarily
-        hmD1betaD1 = Symmetric(M * Diagonal(h) + minus_MD1betaD1)
-        # If the time integration method takes a too large step, h become become negative
-        # letting the factorization fail. In this case the solution is set to `NaN` to force
-        # rejecting the step
-        cholesky!(factorization, hmD1betaD1, check = false)
-        if issuccess(factorization)
-            mul!(tmp1, M, dv)
-            dv[:] = factorization \ tmp1
-        else
-            fill!(dv, NaN)
-        end
+    @trixi_timeit timer() "assemble system matrix" begin
+        system_matrix = assemble_system_matrix!(cache, h, D1, D1_central, equations)
+    end
+    @trixi_timeit timer() "solving elliptic system" begin
+        tmp1 .= dv
+        solve_system_matrix!(dv, system_matrix, tmp1,
```

Comment:
We cannot set `tmp1` in the hyperbolic part immediately because we need `dv` already to be changed for the source terms. I also tried to directly pass `dv` also as third argument to `solve_system_matrix!`, which also works, but seems a bit dangerous to me. Do you think it's save to pass `dv` directly or would you keep it like this or do you have any other idea?

Reply:
It's better to keep it like this. There might be subtle issues assuming that `tmp1` and `dv` do not alias when we switch to `ldiv!`.